### PR TITLE
Review rating

### PR DESCRIPTION
**Pull request by pallavichavda04**

### CODE REVIEW

File: src/api/reviewRating/review_rating.controller.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';
+import { GetUser } from '../users/guard/getUser.guard';
+import { RolesGuard } from '../auth/guard/role.guard';
+import { AuthGuard } from '@nestjs/passport';
+@ApiTags('reviews-ratings')
+@Controller('reviews-ratings')
+export class ReviewRatingController {
+    constructor(private readonly reviewRatingService: ReviewRatingService) { }
+
+    @Post()
```
Comment (yuvraj-bytes) [inline src/api/reviewRating/review_rating.controller.ts:14]: Add Auth Guard.

File: src/api/reviewRating/review_rating.module.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import { ReviewRatingController } from './review_rating.controller';
+import { ReviewRatingService } from './review_rating.service';
+import { MongooseModule } from '@nestjs/mongoose';
+import { MulterModule } from '@nestjs/platform-express';
+import { ReviewRating, ReviewRatingSchema } from './schema/review-rating.schema';
+import { ErrorHandlerService } from 'src/utills/error-handler.service';
+@Module({
+    imports: [
+        MongooseModule.forFeature([{ name: ReviewRating.name, schema: ReviewRatingSchema }]),
+        MulterModule.register({
```
Comment (yuvraj-bytes) [inline src/api/reviewRating/review_rating.module.ts:11]: Remove this. Don't Just Copy Paste.

File: src/api/reviewRating/review_rating.service.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            const createdReviewRating = new this.reviewRatingModel(reviewRatingDto);
+            createdReviewRating.save();
+            return { statusCode: HttpStatus.OK, message: MESSAGE.REVIEW_RATING_ADDED, data: createdReviewRating }
+        }
+        catch (error) {
+            await this.errorHandlerService.HttpException(error);
+        }
+    }
+
+    async getReviewsRatingsByCourseId(courseId: string): Promise<ResponseDto> {
```
Comment (yuvraj-bytes) [inline src/api/reviewRating/review_rating.service.ts:29]: How we will Get the User Name? Make sure you get all the information.

File: src/app.module.ts [hunk 4]
```diff
@@ -19,14 +20,15 @@ import { ThrottlerModule } from '@nestjs/throttler';
     MongooseModule.forRootAsync({
       imports: [ConfigModule],
       useFactory: async (configService: ConfigService) => ({
-        uri: configService.get<string>('MONGODB_URL'),
+        uri: configService.get<string>('DB_URL'),
```
Comment (yuvraj-bytes) [inline src/app.module.ts:23]: Handle this with development and production.

File: src/api/reviewRating/review_rating.service.ts [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+                        'courseId': new mongoose.Types.ObjectId(courseId)
+                    }
+                }, {
+                    '$lookup': {
+                        'from': 'users',
+                        'localField': 'userId',
+                        'foreignField': '_id',
+                        'as': 'userId'
+                    }
+                }, {
```
Comment (yuvraj-bytes) [inline src/api/reviewRating/review_rating.service.ts:65]: add project

**Comment by admsev** [comment]
@gooroodev would you mind to take a look please?

**Comment by gooroodev** [comment]
@admsev, thanks for summoning me!

### 1. Summary of Changes:
The pull request introduces a new feature for handling reviews and ratings for courses. The main changes include:
- Addition of a `ReviewRating` module, controller, service, and schema to manage reviews and ratings.
- Modifications to the `Course` schema to include `raating` and `averageRating` fields.
- Updates to the `dev.env` and `package.json` files for environment configuration and script adjustments.
- Adjustments to imports in several files to correct paths.
- Addition of constants for review and rating messages.
- Minor refactoring in the `app.module.ts` to conditionally load environment variables based on the `NODE_ENV`.

### 2. Issues, Bugs, and Typos:
#### Typo in `Course` Schema:
- **Issue:** Typo in the `Course` schema where `raating` should be `rating`.
  ```diff
  - @Prop()
  - raating: number;
  + @Prop()
  + rating: number;
  ```

#### Incorrect Update Query in `updateReviewRating` Method:
- **Issue:** The `updateReviewRating` method uses `courseId` instead of `reviewRatingId` to find the document to update.
  ```diff
  - const updatedReviewRating = await this.reviewRatingModel.findOneAndUpdate({ courseId: reviewRatingId }, reviewRatingDto, { new: true });
  + const updatedReviewRating = await this.reviewRatingModel.findOneAndUpdate({ _id: reviewRatingId }, reviewRatingDto, { new: true });
  ```

#### Redundant `raating` Field in `Course` Schema:
- **Issue:** The `raating` field seems redundant and is likely a typo. It should be removed.
  ```diff
  - @Prop()
  - raating: number;
  ```

### 3. General Review of Code Quality and Style:
#### Positive Aspects:
- **Modularity:** The code is well-modularized with clear separation of concerns across different files and modules.
- **Use of DTOs:** The use of Data Transfer Objects (DTOs) for request validation is a good practice.
- **Swagger Integration:** The integration with Swagger for API documentation is beneficial.
- **Error Handling:** The use of a centralized error handling service is a good practice.

#### Areas for Improvement:
- **Code Consistency:** Ensure consistent naming conventions across the codebase. For example, `raating` should be `rating`.
- **Comments and Documentation:** Adding comments and documentation within the code can improve readability and maintainability.
- **Testing:** Ensure that there are comprehensive unit and integration tests for the new features introduced.

### Improved Code:
#### Fix Typo in `Course` Schema:
```typescript
import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';

@Schema()
export class Course {
    _id: string;

    @Prop()
    title: string;

    @Prop()
    description: string;

    @Prop()
    price: number;

    @Prop()
    instructor_id: string;

    @Prop()
    image: string;

    @Prop()
    product_id: string;

    @Prop()
    price_id: string;

    @Prop()
    rating: number; // Fixed typo

    @Prop({ required: false })
    averageRating: number;

    @Prop({ default: Date.now })
    created_at: Date;

    @Prop({ default: Date.now })
    updated_at: Date;
}

export const CourseSchema = SchemaFactory.createForClass(Course);
```

#### Correct Update Query in `updateReviewRating` Method:
```typescript
async updateReviewRating(reviewRatingDto: ReviewRatingDto, reviewRatingId: string): Promise<ResponseDto> {
    try {
        const updatedReviewRating = await this.reviewRatingModel.findOneAndUpdate({ _id: reviewRatingId }, reviewRatingDto, { new: true });
        // Update course data with average rating
        const courseId = reviewRatingDto.courseId;
        const averageRatingData = await this.reviewRatingModel.aggregate([
            {
                '$match': {
                    'courseId': new mongoose.Types.ObjectId(courseId)
                }
            }, {
                '$group': {
                    '_id': '$courseId',
                    'averageRating': { '$avg': '$rating' },
                    'totalReviews': { '$sum': 1 }
                }
            }
        ]).exec();
        if (averageRatingData.length > 0) {
            const averageRating = averageRatingData[0].averageRating;
            await this.courseModel.updateOne({ _id: courseId }, { averageRating: averageRating });
        }
        return { statusCode: HttpStatus.OK, message: MESSAGE.REVIEW_RATING_UPDATED, data: updatedReviewRating }
    }
    catch (error) {
        await this.errorHandlerService.HttpException(error);
    }
}
```

### General Review:
The code is generally well-structured and follows good practices for modularity and error handling. The addition of the review and rating feature is a valuable enhancement to the application. Addressing the mentioned issues and typos will further improve the quality and maintainability of the code.

Yours, [Gooroo.dev](https://github.com/apps/gooroo-dev). I'd love a reaction or reply to know what you think.

**Pull request merged**